daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/SuspensionTracker.scala

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.processors
+
+import scala.collection.mutable.Queue
+
+import org.apache.daffodil.dsom.RuntimeSchemaDefinitionError
+import org.apache.daffodil.exceptions.Assert
+
+class SuspensionTracker {
+
+  private val suspensionsYoung = new Queue[Suspension]
+  private val suspensionsOld = new Queue[Suspension]
+
+  private val suspensionWaitYoung = 5
+  private val suspensionWaitOld = 100
+
+  private var count: Int = 0
+
+  def trackSuspension(s: Suspension): Unit = {
+    suspensionsYoung.enqueue(s)
+  }
+
+  /**
+   * Attempts to evaluate suspensions. Old suspensions are evaluated less
+   * frequently than young suspensions. Any young suspensions that fail to
+   * evaluate are moved to the old suspensions list. If we evaluate old
+   * suspensions, we attempt to evaluate them first, with the hope that their
+   * resolution might make the young suspensions more likely to evaluate.
+   */
+  def evalSuspensions(): Unit = {
+    if (count % suspensionWaitOld == 0) {
+      evalSuspensionQueue(suspensionsOld)
+    }
+    if (count % suspensionWaitYoung == 0) {
+      evalSuspensionQueue(suspensionsYoung)
+      while (suspensionsYoung.nonEmpty) {
+        suspensionsOld.enqueue(suspensionsYoung.dequeue)
+      }
+    }
+
+    if (count == suspensionWaitOld) {
+      count = 0 
+    } else {
+      count += 1
+    }
+  }
+
+  /**
+   * Evaluates all suspensions until either they are all evaluated or a
+   * deadlock is detected. This moves all young suspensions to the old queue,
+   * and evaluates all old suspensions. If the old queue is non-empty, that
+   * means some suspensions are blocked, likely due to a circular deadlock, and
+   * we output diagnostics.
+   */
+  def requireFinal(): Unit = {
+    while (suspensionsYoung.nonEmpty) {
+      suspensionsOld.enqueue(suspensionsYoung.dequeue)
+    }
+
+    evalSuspensionQueue(suspensionsOld)
+
+    if (suspensionsOld.length == 1) {
+      Assert.invariantFailed("Single suspended expression making no forward progress. " + suspensionsOld(0))
```

Comment:
Actually if you change this to Assert.invariant(suspensionsOld.length != 1, "Single suspended expression ...") then that would fix the code-cov warning I'm seeing.

Reply:
We don't have an invariant() function that accepts a string. I'll see if the invariant macro can be changed to allow a message that just defaults to empty string.

Reply:
Turns out macros aren't allowed to have default arguments. But function overloading is allowed, and it's easy enough to add a new macro that accepts 2 parameters. Change has been made. This should fix this line of coverage.